### PR TITLE
Retry availability until the receiver of the request is dropped

### PR DESCRIPTION
**Pull request by rphmeier**
and cancel approval work in approval-voting which is relative to a finalized block.

### CODE REVIEW

File: node/network/availability-recovery/src/lib.rs [hunk 1]
```diff
@@ -442,59 +419,99 @@ fn reconstructed_data_matches_root(
...
+
+		// these are reverse order, so remove is fine.
+		for index in indices_to_remove {
+			tracing::debug!(
+				target: LOG_TARGET,
+				candidate_hash = ?self.candidate_hash,
+				"Receiver for available data dropped.",
+			);
+
+			self.awaiting.remove(index);
```
Comment (ordian) [inline node/network/availability-recovery/src/lib.rs:473]: that's `O(n^2)` in the worst case, we can do this in `O(n)` by using `swap_remove`
as we do in https://github.com/paritytech/polkadot/blob/726db12a487c08589a1ef2aa32e03b971a150d04/node/network/bridge/src/validator_discovery.rs#L282-L284

as long as we don't care about the ordering

**Review by ordian — APPROVED** [review]
Looks reasonable.

File: node/network/availability-recovery/src/lib.rs (hunk 1 above)
Comment (rphmeier) [reply node/network/availability-recovery/src/lib.rs:473]: The size here will generally be quite small; usually one, in fact. but I will make this change.

**Comment by rphmeier** [comment]
bot merge

**Comment by joao-paulo-parity** [comment]
https://github.com/paritytech/polkadot/pull/2763#issuecomment-810305669 is likely the same problem as https://github.com/paritytech/parity-processbot/issues/268. I'll be pushing a fix soon.

**Pull request merged**